tests/unit/test_agent_controller.py

```diff
@@ -355,3 +355,71 @@ async def test_step_max_budget_headless(mock_agent, mock_event_stream):
     # In headless mode, throttling results in an error
     assert controller.state.agent_state == AgentState.ERROR
     await controller.close()
+
+
+@pytest.mark.asyncio
+async def test_message_action_user_input_headless(mock_agent, mock_event_stream):
+    # Test with default fake response
+    controller = AgentController(
+        agent=mock_agent,
+        event_stream=mock_event_stream,
+        max_iterations=10,
+        sid='test',
+        confirmation_mode=False,
+        headless_mode=True,
+    )
+    controller.state.agent_state = AgentState.RUNNING
+    message_action = MessageAction(content='Test message', wait_for_response=True)
+    message_action._source = EventSource.AGENT
+    await controller.on_event(message_action)
+    # In headless mode with default fake response, should continue running
+    assert controller.state.agent_state == AgentState.RUNNING
+    mock_event_stream.add_event.assert_called_once()
+    args = mock_event_stream.add_event.call_args[0]
+    assert isinstance(args[0], MessageAction)
+    assert args[0].content == 'continue'
+    await controller.close()
+
+    # Test with custom fake response
+    mock_event_stream.reset_mock()
+    custom_response = 'custom response'
+    controller = AgentController(
+        agent=mock_agent,
+        event_stream=mock_event_stream,
+        max_iterations=10,
+        sid='test',
+        confirmation_mode=False,
+        headless_mode=True,
+        fake_user_response_fn=lambda _: custom_response,
+    )
+    controller.state.agent_state = AgentState.RUNNING
+    message_action = MessageAction(content='Test message', wait_for_response=True)
+    message_action._source = EventSource.AGENT
+    await controller.on_event(message_action)
+    # In headless mode with custom fake response, should continue running
+    assert controller.state.agent_state == AgentState.RUNNING
+    mock_event_stream.add_event.assert_called_once()
+    args = mock_event_stream.add_event.call_args[0]
+    assert isinstance(args[0], MessageAction)
+    assert args[0].content == custom_response
+    await controller.close()
+
+
+@pytest.mark.asyncio
+async def test_message_action_user_input_non_headless(mock_agent, mock_event_stream):
+    controller = AgentController(
+        agent=mock_agent,
+        event_stream=mock_event_stream,
+        max_iterations=10,
+        sid='test',
+        confirmation_mode=False,
+        headless_mode=False,
+    )
+    controller.state.agent_state = AgentState.RUNNING
+    message_action = MessageAction(content='Test message', wait_for_response=True)
+    message_action._source = EventSource.AGENT
+    await controller.on_event(message_action)
+    # In non-headless mode, should wait for user input
+    assert controller.state.agent_state == AgentState.AWAITING_USER_INPUT
+    mock_event_stream.add_event.assert_not_called()
```

Comment:
@openhands-agent This test is failing, and it fails here, at the line asserting that add_event is not called:

> @pytest.mark.asyncio
    async def test_message_action_user_input_non_headless(mock_agent, mock_event_stream):
        controller = AgentController(
            agent=mock_agent,
            event_stream=mock_event_stream,
            max_iterations=10,
            sid='test',
            confirmation_mode=False,
            headless_mode=False,
        )
        controller.state.agent_state = AgentState.RUNNING
        message_action = MessageAction(content='Test message', wait_for_response=True)
        message_action._source = EventSource.AGENT
        await controller.on_event(message_action)
        # In non-headless mode, should wait for user input
        assert controller.state.agent_state == AgentState.AWAITING_USER_INPUT
>       mock_event_stream.add_event.assert_not_called()

> tests/unit/test_agent_controller.py:424: 

> self = <MagicMock name='mock.add_event' id='[140554164855408](tel:140554164855408)'>
> E           AssertionError: Expected 'add_event' to not have been called. Called 1 times.
> E           Calls: [call(AgentStateChangedObservation(content='', agent_state=<AgentState.AWAITING_USER_INPUT: 'awaiting_user_input'>, observation='agent_state_changed'), <EventSource.ENVIRONMENT: 'environment'>)].

Look very carefully at the error. First, ALWAYS when an agent changes state, the app adds an AgentStateChangedObservation. Second, look at the attributes of this AgentStateChangedObservation: it says that the state of the agent is exactly AWAITING_USER_INPUT, which is what we wanted!

Fix the test.

Reply:
@openhands-agent You didn't address this review comment enough, about the unit test.

The test should look at the event and its properties and verify that:
- the event is a AgentStateChangedObservation
- its agent_state attribute is AgentState.AWAITING_USER_INPUT.